### PR TITLE
add the RHICf library (for only storing raw data in BFC level)

### PR DESCRIPTION
**Pull request by ggfdsa10**
1. new 2 makers 
a) StRHICfDbMaker (RHICf calibration data table maker)
b) StRHICfRawHitMaker (storing raw data maker)

2. modified StEvent, StBFChain and add some new RHICf class 
The revised classes are essential for makers to add later. (RHICf calibration and reconstruction makers)

3. new MuRHICf class (for store in MuDst) and some MuDst base class modifying

4. add the StRHICfUtil folder and class (there to be added another reconstruction tool)

### CODE REVIEW

**Comment by starsdong** [comment]
Hi Seunghwan,

Thank you for putting this PR up. Just to clarify how we proceed with the review.

1) These requested updates will go through the designated reviews
     BFChain:   Gene (@genevb)
     StEvent:  Jason (@klendathu2k)
     MuDstMaker:    Daniel B. (@jdbrice)

2) StRHICfDbMaker and StRHICfRawHitMaker are two new makers. We should go through the standard code review process. I will ask a couple of our colleagues to help on this (follow up on emails)

3) StRHICfUtil directory
     I am wondering whether you need this for production (meaning daq->MuDst) or they are used at the analyses level?

Thanks

**Comment by ggfdsa10** [comment]
Hello Xin,

Thank you for your follow-up.
Mostly, StRHICfUtil works at the analysis level (calibration and reconstruction).
But, There is some function (such as the tower size checker in StRHICfFunction in StRHICfUtil folder).
This tiny function (this name is checkGSOBarSize() ) is used in many parts such as StRHICfRawHitMaker, StMuRHICfUtil, etc.

so I would like to add this directory. 

Thank you.

P.S:
Also, StRHICfFunction includes the RHICf run type checker. 
this function is very important for reconstruction.

**Comment by starsdong** [comment]
Thanks to Dmitry A. and Jeff for help review the new makers (StRHICfDbMaker and StRHICfRawHitMaker). Dmitry and Jeff are added to the reviewer list.

**Comment by ggfdsa10** [comment]
I think these PR test errors are caused by StRHICfFunction class not being recognized in StMuDST.
should I separate the function like a checkGSOBarSize() from the BFC level?
If I should do that, I will revise the StRHICfUtil directory is used only for the analysis level (MuDst -> picoDst)

**Comment by plexoos** [comment]
Seunghwan, for test purposes, can you provide commands (and input files) to convert raw and/or simulated data to StEven and muDst formats? Thanks

**Comment by ggfdsa10** [comment]
Hello Dmitry, 
I'm attached runBfc.C file link.
https://github.com/ggfdsa10/STAR-RHICf/blob/main/runBfc.C

"rhicfdat" BFC option is RHICf raw data saving option for convert from .daq to MuDst.
and that code includes one .daq file path. "st_rhicf_18178002_raw_1000004.daq"

Thank you.

**Comment by ggfdsa10** [comment]
Dear All,
I think all of the PR test actions are checkers about the process of BFC -> MuDst level.

As I said before, that error seems caused by StRHICfFunction class in StMuRHICfUtil class.
also, all of that actions do not include "RHICfUtil" lib that is "rhicfutil" BFC option. and that option is used by only RHICf process (that library is generated from the StRHICfUtil directory)

due to people who are not using RHICf do not need to call the RHICfUtil library, separating between StMuDST and StRHICfUtil is necessary 

For this reason, I have removed the StRHICfFunction class header and functions at StMuRHICfUtil class.
But, In the "StRHICfRawHitMaker", exist the StRHICfFunction yet.

I'm not sure if I can upload the Pull Request again myself. 
but I would like to request the revised StMuRHICfUtil class again. 
I'm so sorry if you already reviewed this part.

Thank you

File: StRoot/StBFChain/doc/index.html
Comment (klendathu2k) [inline StRoot/StBFChain/doc/index.html]: StBFChain/doc contains files (such as this one) which are automatically generated at compile time, and should not be added to this pull request.

File: StRoot/StEvent/StRHICfCollection.h
Comment (klendathu2k) [inline StRoot/StEvent/StRHICfCollection.h]: Should update the classdef to `ClassDef(StRHICfCollection, 2)`

File: StRoot/StEvent/StRHICfCollection.h
Comment (klendathu2k) [inline StRoot/StEvent/StRHICfCollection.h]: StEventInfo::mRunId and StEventInfo::mId provide _the_ run and event number for each entry in StEvent.  I am not sure why these are necessary.

File: StRoot/StEvent/StRHICfRawHit.h
Comment (klendathu2k) [inline StRoot/StEvent/StRHICfRawHit.h]: You're mixing root types (Int_t) and standard c++ types (unsigned int).  For IO classes (such as this one) we prefer to use the ROOT types, per STAR coding guidelines.  `unsigned int` should be `UInt_t`

File: StRoot/StEvent/StRHICfRawHit.h
Comment (klendathu2k) [inline StRoot/StEvent/StRHICfRawHit.h]: In terms of size, do you really need a full Int_t here?  Could you fit ADC values into a short?

File: StRoot/StRHICfUtil/StRHICfFunction.h
Comment (klendathu2k) [inline StRoot/StRHICfUtil/StRHICfFunction.h]: This appears to be a utility class, rather than a persistent data class.  The preference is to use standard c++ types (int, float, etc...) rather than the ROOT tyles (Int_t, Float_t, etc...).

Also, because this class is a persistent class (i.e. saved to disk) you do not need the ClassDef statement below.

**Review by klendathu2k — CHANGES_REQUESTED** [review]
I don't see any major issues at this point.  I've noted a few things in the event data model which I would like to see addressed.  After they are attended to I may (or may not) have additional comments.

File: StRoot/StEvent/StRHICfRawHit.h
Comment (ggfdsa10) [reply StRoot/StEvent/StRHICfRawHit.h]: I think UShort_t would be the best, Thanks

File: StRoot/StEvent/StRHICfRawHit.h
Comment (ggfdsa10) [reply StRoot/StEvent/StRHICfRawHit.h]: I think UShort_t would be the best, Thank you

File: StRoot/StRHICfUtil/StRHICfFunction.h
Comment (ggfdsa10) [reply StRoot/StRHICfUtil/StRHICfFunction.h]: This class is a utility class, not a saving class.

**Also, because this class is a persistent class (i.e. saved to disk) you do not need the ClassDef statement below.**

as per your comment, if a class is "a utility class", there is no need for the ClassDef, it's right?

Comment (klendathu2k) [reply StRoot/StRHICfUtil/StRHICfFunction.h]: Correct.  If you do not write it to disk (e.g. it is not part of the MuDst or StEvent) then you do not need to declare the ClassDef.

**Review by dmarkh — APPROVED** [review]
StRHICfDbMaker looks good to me, no changes are required.

File: StRoot/StRHICfUtil/StRHICfFunction.h
Comment (ggfdsa10) [reply StRoot/StRHICfUtil/StRHICfFunction.h]: Thank you for the answer.

**Comment by jml985** [comment]
It appears that I was removed as a reviewer (which is fine with me).   I don't have any opinions or insight into all the failed CE tests.  

I did however take a look at the StRHICfDbMaker and StRHICfHitMaker additions.    I think the Db Maker probably meets the normal standards just fine.   The tables are filled once at the beginning of each run, and if there were more than one run would be memory leaks, but I think this is impossible for the offline chain.  The function for deleting tables is empty.   The only code is the setup in the initRun() and a bunch of helper functions translating the DB data into detector specific items.  I don't see any structural issues there.

In the stRHICfHitMaker codes, the setup logic in lines 56-79 is very complex and the actual logic is very simple.   The actual logic appears to be a simple linear assignment of variables each dependent on previous step:

if(!eventPtr)  return kStFatal;

mRHICfCollection = eventPtr->rhicfCollection();
if(!mRHICfCollection) mRHICfCollection=new StRHICfCollection;
if(!mRHICfCollection) return kStFatal;

mRHICfRawHitColl=mRHICfCollection->rawHitCollection();
if(!mRHICfRawHitColl) return kStFatal;

But the existing is logically unclear, and the setupRHICfCollection() function does nothing except make it less clear.

**Comment by starsdong** [comment]
Hi Jeff, I guess you were probably accidentally removed. I added you back in.
For everyone, Dmitry A. and Jeff L. were asked and agreed to review the new two makers: StRHICfDbMaker and StRHICfRawHitMaker as our normal review procedure for new codes.

**Review by jml985 — CHANGES_REQUESTED** [review]
See my previous comment.   The suggested change is to simplify the logic in StRICHfRawHitMaker lines 56-79.   All three variables must be initialized to proceed, and any problem initializing them causes a fatal error.  This is obscured by the nested if statements.

**Comment by ggfdsa10** [comment]
I'm so sorry for my mistake, I don't know why I removed a reviewer.
I guess there is something wrong when I clicked the re-request button

anyway, I don't know am I understanding your comments.
but, as far as I understood your comments, 
StRHICfRawHitMaker needs to simplify a code. such as your comments:

**if(!eventPtr) return kStFatal;

mRHICfCollection = eventPtr->rhicfCollection();
if(!mRHICfCollection) mRHICfCollection=new StRHICfCollection;
if(!mRHICfCollection) return kStFatal;

mRHICfRawHitColl=mRHICfCollection->rawHitCollection();
if(!mRHICfRawHitColl) return kStFatal;**

Also, that have to get to need the **eventPtr->setRHICfCollection(mRHICfCollection)**
because there is no RHICfCollection at all of the event starting points.
I've already checked when that code is present or not present.

and then, 
as per your other comment, StRHICfDbMaker is well done but should remove unnecessary functions like the deleteArrays().
If I was wrong, please let me know.

Thank you!

**Comment by ggfdsa10** [comment]
oh... I'm very sorry for Jason. when I clicked the re-requested button, removed a reviewer again. :(
I will NOT click this button.

**Review by jml985 — APPROVED** [review]
This update is much more clear, thank you.

**Comment by starsdong** [comment]
Hi All,
The two new makers are reviewed and approved by reviewers. We need to follow up on other changes in this PR too.
@klendathu2k, on the StEvent containers, do you have any further comments/suggestions?
@genevb, any request on the BFChain options?
@jdbrice, could you please take a look at the MuDst changes?
THanks and Best Regards

File: StRoot/StRHICfUtil/StRHICfFunction.h
Comment (genevb) [inline StRoot/StRHICfUtil/StRHICfFunction.h]: Is this line needed if you have above:

    #include "StRoot/StEvent/StEnumerations.h"

?

File: StRoot/StBFChain/BigFullChain.h
Comment (genevb) [inline StRoot/StBFChain/BigFullChain.h]: I don't see codes for StRHICfHitMaker and StRHICfPointMaker. Am I missing something? Are those coming later? If so, I think the chain options should be introduced when those are introduced.

**Comment by klendathu2k** [comment]
Hi Xin,

I see many of my comments resolved.  But there are some memory 
allocations that need
to be cleaned up.

Cheers,
Jason

**Comment by plexoos** [comment]
> I'm attached runBfc.C file link. https://github.com/ggfdsa10/STAR-RHICf/blob/main/runBfc.C
> 
> "rhicfdat" BFC option is RHICf raw data saving option for convert from .daq to MuDst. and that code includes one .daq file path. "st_rhicf_18178002_raw_1000004.daq"

The daq file and the chain options have been added to CI tests. See #476 I would recommend to incorporate this along with modifications to StEvent and StRHICfRawHitMaker because that's what `rhicfdat` option depends on.

File: StRoot/StRHICfUtil/StRHICfFunction.h
Comment (ggfdsa10) [reply StRoot/StRHICfUtil/StRHICfFunction.h]: Hello Gene, 

There are some functions that check whether there is correct input data or not. (this is just checking the size of the data)
So some parameters like kRHICfNtower, kRHICfNlayer, and so on are in this source file.

If someone uses the RHICf reconstruction tool without changing something, these checking functions will be meaningless.

If you think not needed these checking functions, then I will remove these functions and StEnumerations.hh include line.

Thank you

File: StRoot/StBFChain/BigFullChain.h
Comment (ggfdsa10) [reply StRoot/StBFChain/BigFullChain.h]: Yes, StRHICfHitMaker and StRHICfPointMaker will be added later. (But these makers have already been made)
I will make new Pull-requests for calibration and reconstruction makers, including these BFC options later.

**Comment by ggfdsa10** [comment]
Hello Jason, 
Please tell me if there is anything I need to revise the code to arrange the parts of the allocation memory.
Thank you

**Comment by ggfdsa10** [comment]
Hello Dmitri,
I don't understand your comments.
Do you mean, I just need to add this macro https://github.com/ggfdsa10/STAR-RHICf/blob/main/runBfc.C in this pull request?

If I don't need to additional revise some BFC options, I will solve Gene's request as below. 

Gene: "I don't see codes for StRHICfHitMaker and StRHICfPointMaker. Am I missing something? Are those coming later? If so, I think the chain options should be introduced when those are introduced."

Please let me know if I'm wrong.
Thank you

File: StRoot/StRHICfUtil/StRHICfFunction.h
Comment (genevb) [reply StRoot/StRHICfUtil/StRHICfFunction.h]: Thanks for your reply, but I think I was not clear. I am saying that _if_ you have the `#include` statement, you _do_not_need_ the `class StEnumerations;` statement because the StEnumerations class is already known from the `#include` statement. You can delete the `class StEnumerations;` line.

**Comment by plexoos** [comment]
> I don't understand your comments.
> Do you mean, I just need to add this macro https://github.com/ggfdsa10/STAR-RHICf/blob/main/runBfc.C in this pull request?

No, you don't need to add your macro. In order to enable that test, you should add the changes from #476 (for example, by cherry-picking the commit) onto this PR branch. If you don't know how to do this, let me know and I'll do it.

Your test, as it is now, will simply make calls to StRHICfRawHitMaker and hopefully fill StEvent with RHICf data. If, however, you have a different macro that somehow verifies the content of the produced output files, I'd be happy to help you to accommodate it as a CI test.

> If I don't need to additional revise some BFC options, I will solve Gene's request as below.
> 
> Gene: "I don't see codes for StRHICfHitMaker and StRHICfPointMaker. Am I missing something? Are those coming later? If so, I think the chain options should be introduced when those are introduced." 

Yes, I agree with Gene, the options calling the non-existent makers should be removed from this PR and introduced along with those new makers in a separate PR. I would add that introducing tests for the new options is highly desirable.

File: StRoot/StRHICfUtil/StRHICfFunction.h
Comment (genevb) [reply StRoot/StRHICfUtil/StRHICfFunction.h]: And there is more...

I do not see the use of any functionality defined in `StEnumerations.h` here in `StRHICfFunctions.h`. So one question to ask is, "Why have the `#include` for `StEnumerations.h` inside `StRHICfFunctions.h`?"

You should answer this question.

One possible answer is that every other code that includes `StRHICfFunctions.h` also needs functions defined in `StEnumerations.h`, so this is a shortcut to avoid having to include `StEnumerations.h` in those other codes.

Another possible answer is that it may be better to change the codes to include `StEnumerations.h` only in those other codes that really need it, and remove both the `class` and the `#include` for `StEnumerations` from `StRHICfFunctions.h`.

I hope that is clear, but please let me know if I need to explain the question better.

Thanks,
-Gene

Comment (ggfdsa10) [reply StRoot/StRHICfUtil/StRHICfFunction.h]: class StEnumerations; this line is my mistake. sorry that is ok if removed.
I will remove it.

and I mean, some constant values for RHICf in the **StEnumerations.h** are used in the **StRHICfFunctions.cxx** source file.
In the **StRHICfFunctions.cxx**, you can see some values like a kRHICfNtower, kRHICfNlayer, etc.. 

the **StRHICfFunction.h** class is called in other classes such as **StRHICfHitMaker**, and so on.
so as you say, the **StEnumeration** class should be better to call only one time using the **StRHICfFunction** in the RHICf library
Likewise, I will remove including part for **StEnumeration.h** in other classes that already inherited **StRHICfFunction** 

Thank you for your comments.

**Comment by ggfdsa10** [comment]
Dear Dmitri,

Now I'm clear about what are you saying. 
I don't know how to add a CI test to this request. 
If you help me with this part I appreciate you.

Also, I share some macro file that is verifying for output MuDst file with you.
https://github.com/ggfdsa10/STAR-RHICf/blob/main/openMuDst.C

If this macro is something wrong or does not function, please tell me 
Also, if there's anything I need to do, please let me know anytime.

Thank you!

**Comment by plexoos** [comment]
Hi Seunghwan,

I pushed the change enabling the CI test to this PR branch.
Please note that your code does not compile due to the following errors:

```
#10 2398.2 g++ -m64 -fPIC -pipe -Wall -Woverloaded-virtual -std=c++0x -Wno-long-long -pthread -Wno-deprecated-declarations -Werror -O2 -g -falign-loops -falign-jumps -falign-functions -Dsl79_gcc485 -D__ROOT__ -DNEW_DAQ_READER -I. -IStRoot/StRHICfRawHitMaker -IStRoot/RTS/src -IStRoot/RTS/include -IStRoot/RTS/trg/include -I. -IStRoot -I.sl79_gcc485/include -I/opt/software/linux-scientific7-x86_64/gcc-4.8.5/root-5.34.38-l3v6vso6qgojm4l2ctwjojs6trbt4hpn/include -c .sl79_gcc485/OBJ/StRoot/StRHICfRawHitMaker/StRHICfRawHitMaker.cxx -o .sl79_gcc485/OBJ/StRoot/StRHICfRawHitMaker/StRHICfRawHitMaker.o
#10 2399.1 .sl79_gcc485/OBJ/StRoot/StRHICfRawHitMaker/StRHICfRawHitMaker.cxx: In member function 'virtual Int_t StRHICfRawHitMaker::Make()':
#10 2399.1 .sl79_gcc485/OBJ/StRoot/StRHICfRawHitMaker/StRHICfRawHitMaker.cxx:70:23: error: 'class StRHICfCollection' has no member named 'setRunNumber'
#10 2399.1    mRHICfCollection -> setRunNumber(gendata[0]);
#10 2399.1                        ^
#10 2399.1 .sl79_gcc485/OBJ/StRoot/StRHICfRawHitMaker/StRHICfRawHitMaker.cxx:71:23: error: 'class StRHICfCollection' has no member named 'setEventNumber'
#10 2399.1    mRHICfCollection -> setEventNumber(gendata[1]);
#10 2399.1                        ^
```

**Comment by ggfdsa10** [comment]
Ok that is revised Thank you

File: StRoot/StRHICfUtil/StRHICfFunction.h
Comment (ggfdsa10) [reply StRoot/StRHICfUtil/StRHICfFunction.h]: I had committed the code. Thank you.

**Comment by ggfdsa10** [comment]
Dear All,

I was wondering if you could tell me when we can receive the code review approval and merge this request if you get a chance.
After this review, We would like to produce all of the raw data to muDst.
Please let me know if there is anything I need to do to review.

Thank you for your convenience,
Seunghwan

File: StRoot/StEvent/StRHICfHit.h
Comment (plexoos) [inline StRoot/StEvent/StRHICfHit.h]: That is an addition of 18 arrays some of which are 3 and 4 dimensional. Have you checked by how much the size of event.root files can increase on average?  These are static arrays so, if I think about this right all production chains will be affected, not just RHICf streams (if there is such thing) On the other hand, if they are empty the automatic ROOT compression should minimize the effect

File: StRoot/StRHICfDbMaker/StRHICfDbMaker.h
Comment (plexoos) [inline StRoot/StRHICfDbMaker/StRHICfDbMaker.h]: It is a maker. We don't version makers for IO

```suggestion
	ClassDef(StRHICfDbMaker, 0)
```

**Review by plexoos — COMMENTED** [review]
Hi Seunghwan,

Sorry for the delay, but unfortunately it is something you can expect when you submit so many changes and include many new files in various packages. Usually making smaller changes more often is more efficient than dumping everything at once. It also helps to target specific packages/maintainers. For example, if you submitted the StRHICfDbMaker in a separate PR it could have been merged as I see that the maintainers already responded and approved. It is not late to do so, for instance, with the MuDst code.

Another general comment is that you should reformat the code to match the STAR coding style guidelines. In particular using tabs for indentation is not allowed. The easiest way to do this would be to run clang-format over the new files.

File: StRoot/StEvent/StRHICfHit.h
Comment (ggfdsa10) [reply StRoot/StEvent/StRHICfHit.h]: Hello Dmitri,
Thank you for your reply, and I understand about this delayed PR Thank you.

We including the RHICf collaborator and I already had a discussion about these variable arrays.
Actually, In the reconstruction process of the RHICf library, 
If some events not satisfying a condition (for example, there is no worth for reconstruction or no have a tower hit),
then We don't store these arrays (except mPlateE, mGSOBarSmallE, mGSOBarLargeE, mL20, and mL90)
But our collaborators told me these arrays, all of the various arrays might be needed.

In spite of that, I know what do you concern about.
As I tell you above, these arrays only store a few events per 100 events. 
(But if there is a satisfying saving condition, all of the arrays are stored fully.)
So I think I have to change these array formats from static array to vector.

Please let me know if these changes are inappropriate.

Thank you,
Seunghwan

File: StRoot/StRHICfDbMaker/StRHICfDbMaker.h
Comment (ggfdsa10) [reply StRoot/StRHICfDbMaker/StRHICfDbMaker.h]: Thank you for pointing that out.

**Comment by ggfdsa10** [comment]
Hello Dmitri, 

Thank you for your comments and sorry for I didn't follow the STAR guidelines.
I think all of the files in this PR have to revise part of the indentation. 
Also, In my PR files, Is there any other STAR coding guideline that I'm missing?
I would appreciate it if you could let me know.

Seunghwan

**Comment by ggfdsa10** [comment]
Dear All, 

I'm sorry for being late.
now I modified the indentation of all code and n-dimension data arrays in StRHICfHit, StMuRHICfHit according to comments.

These commits were designed more can be reduced memory.
also, when someone wants detailed n-dimension data in StRHICfHit, he can obtain that immediately.

Please review these codes and the remaining codes if you get a chance.

Seunghwan

File: StRoot/StRHICfUtil/StRHICfFunction.cxx
Comment (plexoos) [inline StRoot/StRHICfUtil/StRHICfFunction.cxx]: Why this change?
In C++ you probably want to zero C-arrays like this:

```suggestion
std::fill(mLeakageNum, mLeakageNum+3, 0);
std::fill(mValueForRecoNum, mValueForRecoNum+5, 0);
```

**Review by plexoos — APPROVED** [review]
It'd be nice to verify the performance with tests but if there is no intention to add them it's fine by me

File: StRoot/StRHICfUtil/StRHICfFunction.cxx
Comment (ggfdsa10) [reply StRoot/StRHICfUtil/StRHICfFunction.cxx]: Thank you for the suggestion.
I thought FOR loop was more effective.
I commit again

**Comment by plexoos** [comment]
If there are no further comments, is it okay to merge this PR?

**Comment by ggfdsa10** [comment]
Thank you for merging! 
However, there seems to be a problem with the CI test docker.
Please let me know if there is anything I can do to fix this problem.

Sincerely,
Seunghwan

**Comment by plexoos** [comment]
Seunghwan, there was a problem with GitHub package repository access. They seem have fixed the problem and I restarted the jobs

**Comment by ggfdsa10** [comment]
Dear All, 

New change is commit for solve the last change requested one.
I'm so sorry for the inconvenience caused by committing after the commit request.

Seunghwan.

**Comment by plexoos** [comment]
Seunghwan, just create a new PR with fixes. Thanks

**Comment by ggfdsa10** [comment]
Dear All, 

I created a new pull request for Dmitry's request. #511 
I'm very sorry for the inconvenience.

Seunghwan

**Comment by plexoos** [comment]
Oh, so this one was not really merged... Ok, seems like another chance to review the code in #511

**Pull request closed without merging**
